### PR TITLE
Upgrade to Scalameta v4.0.0-M4

### LINKED ISSUE #749: Upgrade to scalameta 4.0.0-M4
The issue #748 suggests there might be thread unsafety in metacp in this line here https://github.com/scalameta/scalameta/blob/3c6e67b341ee4952ce2966b9aefd0ad3d5bdd9ee/semanticdb/metacp/src/main/scala/scala/meta/internal/scalacp/Scalacp.scala#L120

In Scalameta v4.0.0-M2 there is no longer any global mutable map so this might be fixed by upgrading.

### LINKED ISSUE #757: Support -P:semanticdb:text:off
In #756 we upgrade to Scalameta v4.0.0-M4 but still require the non-default `text:on` setting. We should refactor the command-line interface and testkit to work with `text:off`.

### PR DESCRIPTION
**Pull request by olafurpg**
Fixes #749 
Fixes #757 

- move `scala.meta.{Symbol,Signature}` to `scalafix.v0` package
- improve code quality in the type printer:
  - much cleaner pattern matching thanks to new sealed oneof encoding
  - existential types and universal types no longer require hacks because of `Scope`, thanks @xeno-by ❤️ 
- sbt-scalafix is removed in this PR because it will be moved into a separate repo and I prefer to handle the upgrade after the split.

### CODE REVIEW

File: project/ScalafixBuild.scala [hunk 1]
```diff
@@ -74,7 +74,9 @@ object ScalafixBuild extends AutoPlugin with GhpagesKeys {
     lazy val semanticdbSettings = Seq(
       scalacOptions ++= List(
         "-Yrangepos",
-        "-Xplugin-require:semanticdb"
+        "-Xplugin-require:semanticdb",
+        "-P:semanticdb:synthetics:on",
+        "-P:semanticdb:text:on"
```
Comment (xeno-by) [inline project/ScalafixBuild.scala]: Is `-P:semanticdb:text` still required?

File: project/ScalafixBuild.scala [hunk 2]
```diff
@@ -74,7 +74,9 @@ object ScalafixBuild extends AutoPlugin with GhpagesKeys {
     lazy val semanticdbSettings = Seq(
       scalacOptions ++= List(
         "-Yrangepos",
-        "-Xplugin-require:semanticdb"
+        "-Xplugin-require:semanticdb",
+        "-P:semanticdb:synthetics:on",
```
Comment (xeno-by) [inline project/ScalafixBuild.scala]: What functionality requires `-P:semanticdb:synthetics`?

File: scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala [hunk 3]
```diff
@@ -62,9 +61,17 @@ class PrettyType private (table: SymbolTable, shorten: QualifyStrategy) {
...
+  ) {
+    def owner: String = info.symbol.owner
+    def tpe: s.Type = {
+      info.signature match {
+        case s.ValueSignature(tpe) =>
+          tpe
+        case _ =>
+          throw new IllegalArgumentException(info.toProtoString)
+      }
+    }
```
Comment (xeno-by) [inline scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala]: I wonder where this comes into play?

File: scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala [hunk 4]
```diff
@@ -81,11 +88,20 @@ class PrettyType private (table: SymbolTable, shorten: QualifyStrategy) {
...
+        f: PartialFunction[s.SymbolInformation, Iterable[T]]
+    ): List[T] =
       syms.map(info).collect(f).flatten.toList
     def smap[T](f: s.SymbolInformation => T): List[T] =
       syms.map(info).map(f).toList
   }
+  private implicit class XtensionScopeHardlinks(scope: s.Scope) {
+    def infos: List[s.SymbolInformation] =
+      if (scope.hardlinks.isEmpty) scope.symlinks.iterator.map(info).toList
+      else scope.hardlinks.toList
```
Comment (xeno-by) [inline scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala]: Don't we already have a method like that in `semanticdb`?

File: scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala [hunk 5]
```diff
@@ -185,32 +207,24 @@ class PrettyType private (table: SymbolTable, shorten: QualifyStrategy) {
...
           }
-        case t.CLASS_INFO_TYPE =>
-          val classInfo = info.typ.classInfoType.get
+        case s.ClassSignature(Some(tparams), parents, self, Some(decls)) =>
           val declarations =
-            classInfo.declarations.iterator.filter(sym =>
-              !sym.contains("$anon"))
-          val typeParameters = classInfo.typeParameters
-          val parents = classInfo.parents
+            decls.infos.filter(info => !info.symbol.contains("$anon"))
```
Comment (xeno-by) [inline scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala]: Is this filter still necessary after all the fixes that we did around 4.0.0-M2?

File: scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala [hunk 6]
```diff
@@ -185,32 +207,24 @@ class PrettyType private (table: SymbolTable, shorten: QualifyStrategy) {
...
           val isCaseClass = info.is(p.CASE)
-          def objectDecls =
+          def objectDecls: List[Stat] =
             declarations
-              .filter { sym =>
+              .filter { info =>
                 // drop inaccessible ctor due to https://github.com/scalameta/scalameta/issues/1493
-                !sym.endsWith("`<init>`().")
+                !info.symbol.endsWith("`<init>`().")
```
Comment (xeno-by) [inline scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala]: Ditto.

File: scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala [hunk 7]
```diff
@@ -235,17 +249,16 @@ class PrettyType private (table: SymbolTable, shorten: QualifyStrategy) {
...
                   Nil,
                   inits,
                   Self(Name(""), None),
-                  declarations.sflatcollect {
-                    case i
-                        if !i.kind.isConstructor &&
-                          !i.isVarSetter =>
-                      toStat(i)
+                  declarations.flatMap { i =>
+                    if (!i.kind.isConstructor &&
```
Comment (xeno-by) [inline scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala]: Ditto.

File: scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala [hunk 8]
```diff
@@ -273,66 +286,54 @@ class PrettyType private (table: SymbolTable, shorten: QualifyStrategy) {
...
-                      Ctor.Primary(Nil, Name(""), Nil)
-                  }
-                case _ =>
+              val ctor: Ctor.Primary = declarations
+                .collectFirst {
+                  case i if i.kind.isConstructor && i.is(p.PRIMARY) =>
+                    toTree(i) match {
+                      case ctor @ Ctor.Primary(_, _, Nil :: Nil)
+                          if !info.is(p.CASE) =>
+                        ctor.copy(paramss = Nil)
```
Comment (xeno-by) [inline scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala]: Why is this necessary?

File: scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala [hunk 9]
```diff
@@ -273,66 +286,54 @@ class PrettyType private (table: SymbolTable, shorten: QualifyStrategy) {
...
-                  declarations.sflatcollect {
-                    case i
-                        if !i.kind.isConstructor &&
-                          !i.kind.isField &&
-                          !i.isVarSetter &&
-                          !isSyntheticMember(i) =>
-                      toStat(i)
+                  declarations.flatMap { i =>
+                    if (!i.kind.isConstructor &&
+                      !i.kind.isField &&
```
Comment (xeno-by) [inline scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala]: No longer possible.

File: scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala [hunk 10]
```diff
@@ -347,29 +348,23 @@ class PrettyType private (table: SymbolTable, shorten: QualifyStrategy) {
...
+    case s.RepeatedType(targ) =>
+      // Workaround for https://github.com/scalameta/scalameta/issues/1497
+      s.TypeRef(
+        prefix = s.NoType,
+        symbol = "scala.collection.Seq#",
+        typeArguments = targ :: Nil
+      )
+    case targ =>
+      targ
+  }
```
Comment (xeno-by) [inline scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala]: So this is where this ticket comes from!

File: scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala [hunk 11]
```diff
@@ -381,10 +376,10 @@ class PrettyType private (table: SymbolTable, shorten: QualifyStrategy) {
...
-  def toTypeBounds(lo: Option[s.Type], hi: Option[s.Type]): Type.Bounds =
+  def toTypeBounds(lo: s.Type, hi: s.Type): Type.Bounds =
     Type.Bounds(
-      lo.filterNot(TypeExtractors.Nothing.matches).map(toType),
-      hi.filterNot(TypeExtractors.Any.matches).map(toType)
+      Some(lo).filterNot(TypeExtractors.Nothing.matches).map(toType),
+      Some(hi).filterNot(TypeExtractors.Any.matches).map(toType)
     )
 
   def toStat(info: s.SymbolInformation): Option[Stat] = {
```
Comment (xeno-by) [inline scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala]: Is the next line still necessary?

File: scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala [hunk 12]
```diff
@@ -396,9 +391,12 @@ class PrettyType private (table: SymbolTable, shorten: QualifyStrategy) {
       try {
         Some(toTree(info).asInstanceOf[Stat])
       } catch {
-        case e: NoSuchElementException =>
+        case _: NoSuchElementException =>
           None
         case NonFatal(e) =>
+          e.setStackTrace(e.getStackTrace.filter { e =>
+            e.getClassName.startsWith("scalafix")
+          })
```
Comment (xeno-by) [inline scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala]: Why is this necessary?

File: scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala [hunk 13]
```diff
@@ -513,18 +500,17 @@ class PrettyType private (table: SymbolTable, shorten: QualifyStrategy) {
...
-            case _ =>
+            case s.NoType =>
               if (shorten.isName) {
                 name
               } else {
                 toTypeRef(info(symbol))
               }
+            case _: s.SingleType | _: s.ThisType | _: s.SuperType |
+                _: s.ConstantType =>
+              Type.Select(toTermRef(prefix), name)
```
Comment (xeno-by) [inline scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala]: Can a constant type be a prefix to anything?

File: scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala [hunk 14]
```diff
@@ -534,84 +520,71 @@ class PrettyType private (table: SymbolTable, shorten: QualifyStrategy) {
...
-      val existential = tpe.existentialType.get
-      withPlaceholders(existential.typeParameters) { () =>
-        toType(existential.tpe.get)
+    case s.ThisType(symbol) =>
+      Type.Select(
+        Term.This(Name.Anonymous()),
+        Type.Name(this.info(symbol).name)
+      )
+    case s.SuperType(prefix @ _, symbol) =>
+      // TODO: prefix
```
Comment (xeno-by) [inline scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala]: FIXME with a ticket?

File: scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala [hunk 15]
```diff
@@ -534,84 +520,71 @@ class PrettyType private (table: SymbolTable, shorten: QualifyStrategy) {
...
+      Type.Select(
+        Term.Super(Name.Anonymous(), Name.Anonymous()),
+        Type.Name(this.info(symbol).name)
+      )
+    case s.ConstantType(_) =>
+      toType(tpe.widen)
+    case s.ExistentialType(underlying, Some(declarations)) =>
+      withHardlinks(declarations.hardlinks) { () =>
```
Comment (xeno-by) [inline scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala]: How did you find dealing with hardlinks?

File: scalafix-core/shared/src/main/scala/scalafix/v0/Symbol.scala [hunk 16]
```diff
@@ -0,0 +1,152 @@
+package scalafix.v0
+
+import scala.compat.Platform.EOL
+import scala.meta.internal.semanticdb.Scala._
+import scalafix.internal.util.SymbolOps.Root
+
+sealed trait Symbol extends Product {
```
Comment (xeno-by) [inline scalafix-core/shared/src/main/scala/scalafix/v0/Symbol.scala]: Oh hey. Long time no see :)

File: scalafix-tests/unit/src/test/scala/scalafix/tests/core/PrettyTypeSuite.scala [hunk 17]
```diff
@@ -47,13 +46,13 @@ class PrettyTypeSuite extends BasePrettyTypeSuite {
 }
 
 // This test is slow, no need to run it on every PR
-@Ignore
+//@Ignore
```
Comment (xeno-by) [inline scalafix-tests/unit/src/test/scala/scalafix/tests/core/PrettyTypeSuite.scala]: Is this comment accidental?

File: scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala (hunk 5 above)
Comment (olafurpg) [reply scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala]: Good question. Doesn't seem required anymore, I removed it.

File: scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala (hunk 6 above)
Comment (olafurpg) [reply scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala]: This is fixed.

File: scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala (hunk 8 above)
Comment (olafurpg) [reply scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala]: To produce the syntax `class Foo` instead of `class Foo()`. I added a comment to document the reason.

File: scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala (hunk 9 above)
Comment (olafurpg) [reply scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala]: Removed.

File: scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala (hunk 10 above)
Comment (olafurpg) [reply scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala]: Yup. `RepeatedType` can appear in both `Init` position and also for `Defn.Def.decltpe`

```scala
case class A(a: AnyRef*)
```
currently produces a `val a: AnyRef*` method symbol

File: scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala (hunk 12 above)
Comment (olafurpg) [reply scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala]: I've replaced this with a `fatalErrors: Boolean` option to control if exceptions are caught or not. It's possible to set `ExplicitResultTypes.fatalWarnings = true` to control the option. There are still too many NoSuchElementException, for example

- https://github.com/scalameta/scalameta/issues/1625
- https://github.com/scalameta/scalameta/issues/1491
- https://github.com/scalameta/scalameta/issues/1504

File: scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala (hunk 13 above)
Comment (olafurpg) [reply scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala]: I don't think so, removing it seems to be OK. This was a direct refactoring of the old implementation. Hurray for breaking down `SingletonType` 🎉

File: scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala (hunk 14 above)
Comment (olafurpg) [reply scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala]: Done.

File: scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala (hunk 15 above)
Comment (olafurpg) [reply scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala]: It's been great so far 😊 Removed a lot of `???`/WTFs in the old implementation. I currently keep a mutable scope for hardlink lookups, I could make it immutable but that would require adding a `(hardlinks: Map[String, SymbolInformation)` parameter everywhere for questionable benefit.

File: scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala (hunk 4 above)
Comment (xeno-by) [reply scalafix-core/shared/src/main/scala/scalafix/internal/util/PrettyType.scala]: Ah, I missed the `.map(info)` part.

**Comment by xeno-by** [comment]
LGTM.
